Oracle RMAN hangs when using Minio for Azure

1. Created Minio application service using the steps described here https://open.microsoft.com/2017/11/09/s3cmd-amazon-s3-compatible-apps-azure-storage/

```
$ az appservice plan create \
    --name "MinioAppPlan" \
    --is-linux \
    --sku B1 \
    --resource-group "mari184V2" \
    --location "CentralUS"
```

```
$ az webapp create \
    --name "aleminio" \
    --deployment-container-image-name "minio/minio" \
    --plan "MinioAppPlan" \
    --resource-group "mari184V2"
```
	
```
$ az webapp config appsettings set \
    --settings "MINIO_ACCESS_KEY=..." "MINIO_SECRET_KEY=..." "PORT=9000" \
    --name "aleminio" \
    --resource-group "mari184V2"
```

```
$ az webapp config set \
    --startup-file "gateway azure" \
    --name "aleminio" \
    --resource-group "mari184V2"
```
	
2. Installed Oracle Secure Backup module using the following steps: https://blog.pythian.com/backup-oracle-databases-aws-s3/

`$ /u01/app/oracle/product/18.3.0/dbhome_1/jdk/bin/java -jar osbws_install.jar -AWSID ... -AWSKey ... -otnUser mariam.kupa@gmail.com     -walletDir $ORACLE_HOME/dbs/osbws_wallet     -libDir $ORACLE_HOME/lib -awsEndPoint aleminio.azurewebsites.net  -location maribucket
`
The same module works with AWS and I am able to back up the database. But it hangs when using Minio.

RMAN script that is working with AWS but hangs with Minio is the following:

```
run{
  allocate channel c1 device type sbt parms='SBT_LIBRARY=libosbws.so,SBT_PARMS=(OSB_WS_PFILE=/u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbwsorcl1.ora)';
  backup current controlfile;
  }
```
  
Where /u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbwsorcl1.ora file contains:

```
OSB_WS_HOST=http://aleminio.azurewebsites.net
#OSB_WS_LOCATION=maribucket
OSB_WS_BUCKET=maribucket
OSB_WS_VIRTUAL_HOST=FALSE
OSB_WS_WALLET='location=file:/u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbws_wallet CREDENTIAL_ALIAS=_aws'
```

After some time the following error is returned by RMAN:

```
ORA-27192: skgfcls: sbtclose2 returned error - failed to close file
ORA-19511: non RMAN, but media manager or vendor specific failure, error text:
   LHT-00027: Key not found in hash table.
LHT-00027: Key not found in hash table.
LHT-00027: Key not found in hash table.
```

During RMAN backup I have noticed that metadata files are created in the Azure container through Minio, moreover from database I see that RMAN reads datafile but is not able to write backup into Azure container. 

How can I trace what happens on Minio application service, why it hangs?

@kannappanr Thank you!

Good news is that if I do by the following way:

```
# ./minio  gateway azure
...
Endpoint: ...  http://127.0.0.1:9000 ...
```

and then change osbwsorcl1.ora file that is used by RMAN for contacting media library:

> #OSB_WS_HOST=http://aleminio.azurewebsites.net
> **OSB_WS_HOST=http://127.0.0.1:9000**
> #OSB_WS_LOCATION=maribucket
> OSB_WS_BUCKET=maribucket
> OSB_WS_VIRTUAL_HOST=FALSE
> OSB_WS_WALLET='location=file:/u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbws_wallet CREDENTIAL_ALIAS=_aws'

Then it works:

```
RMAN> run{
  allocate channel c1 device type sbt parms='SBT_LIBRARY=libosbws.so,SBT_PARMS=(OSB_WS_PFILE=/u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbwsorcl1.ora)';
  backup current controlfile;
  }2> 3> 4> 

allocated channel: c1
channel c1: SID=153 instance=orcl1 device type=SBT_TAPE
channel c1: Oracle Secure Backup Web Services Library VER=12.2.0.2
...
Finished backup at 04-DEC-18
```
Which is great of course!!

But before this step there is one more step to install this media libary(2nd step in my initial post) that works when using Minio App Service and does not work with Minio binary:

#Minio App Service

```
[oracle@rac1 swtmp]$ /u01/app/oracle/product/18.3.0/dbhome_1/jdk/bin/java -jar osbws_install.jar     -AWSID ...     -AWSKey ...     -otnUser mari@gmail.com     -walletDir $ORACLE_HOME/dbs/osbws_wallet     -libDir $ORACLE_HOME/lib -awsEndPoint aleminio.azurewebsites.net  -location maribucket
Oracle Secure Backup Web Service Install Tool, build 12.2.0.1.0DBBKPCSBP_2018-06-12
AWS credentials are valid.
Oracle Secure Backup Web Service wallet created in directory /u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbws_wallet.
Oracle Secure Backup Web Service initialization file /u01/app/oracle/product/18.3.0/dbhome_1/dbs/osbwsorcl1.ora created.
Downloading Oracle Secure Backup Web Service Software Library from file osbws_linux64.zip.
Download complete.
```

##Minio binary

First try using **-awsEndPoint http://127.0.0.1 -awsPort 9000** option:

```
[oracle@rac1 swtmp]$ /u01/app/oracle/product/18.3.0/dbhome_1/jdk/bin/java -jar osbws_install.jar     -AWSID ...     -AWSKey ...    -otnUser mari@gmail.com     -walletDir $ORACLE_HOME/dbs/osbws_wallet     -libDir $ORACLE_HOME/lib -awsEndPoint http://127.0.0.1 -awsPort 9000  -location maribucket
Oracle Secure Backup Web Service Install Tool, build 12.2.0.1.0DBBKPCSBP_2018-06-12
Exception in thread "main" java.lang.RuntimeException: java.lang.IllegalArgumentException: URI can't be null.
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1506)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
	at sun.net.www.protocol.http.HttpURLConnection.getHeaderField(HttpURLConnection.java:3036)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:489)
	at com.amazon.s3.ListBucketResponse.<init>(ListBucketResponse.java:113)
	at com.amazon.s3.AWSAuthConnection.listBucket(AWSAuthConnection.java:283)
	at com.amazon.s3.AWSAuthConnection.listBucket(AWSAuthConnection.java:261)
	at com.amazon.s3.AWSAuthConnection.connect(AWSAuthConnection.java:825)
	at oracle.backup.s3.install.S3Config.s3Connect(S3Config.java:447)
	at oracle.backup.s3.install.S3Config.doConfig(S3Config.java:228)
	at oracle.backup.s3.install.S3Config.main(S3Config.java:212)
Caused by: java.lang.IllegalArgumentException: URI can't be null.
	at sun.net.spi.DefaultProxySelector.select(DefaultProxySelector.java:148)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1150)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050)
	at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:984)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1564)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
	... 7 more
```
The second try with  **-awsEndPoint http://127.0.0.1:9000** 

```
[oracle@rac1 swtmp]$ /u01/app/oracle/product/18.3.0/dbhome_1/jdk/bin/java -jar osbws_install.jar     -AWSID ...     -AWSKey ...     -otnUser mari@gmail.com     -walletDir $ORACLE_HOME/dbs/osbws_wallet     -libDir $ORACLE_HOME/lib -awsEndPoint http://127.0.0.1:9000  -location maribucket
Oracle Secure Backup Web Service Install Tool, build 12.2.0.1.0DBBKPCSBP_2018-06-12
Exception in thread "main" java.lang.RuntimeException: java.lang.IllegalArgumentException: URI can't be null.
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1506)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
	at sun.net.www.protocol.http.HttpURLConnection.getHeaderField(HttpURLConnection.java:3036)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:489)
	at com.amazon.s3.ListBucketResponse.<init>(ListBucketResponse.java:113)
	at com.amazon.s3.AWSAuthConnection.listBucket(AWSAuthConnection.java:283)
	at com.amazon.s3.AWSAuthConnection.listBucket(AWSAuthConnection.java:261)
	at com.amazon.s3.AWSAuthConnection.connect(AWSAuthConnection.java:825)
	at oracle.backup.s3.install.S3Config.s3Connect(S3Config.java:447)
	at oracle.backup.s3.install.S3Config.doConfig(S3Config.java:228)
	at oracle.backup.s3.install.S3Config.main(S3Config.java:212)
Caused by: java.lang.IllegalArgumentException: URI can't be null.
	at sun.net.spi.DefaultProxySelector.select(DefaultProxySelector.java:148)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1150)
	at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:1050)
	at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:984)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1564)
	at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1492)
	at java.net.HttpURLConnection.getResponseCode(HttpURLConnection.java:480)
	... 7 more
```

--- a/cmd/utils.go
+++ b/cmd/utils.go
@@ -105,6 +105,12 @@ const (
 	httpsScheme = "https"
 )
 
+// nopCharsetConverter is a dummy charset convert which just copies input to output,
+// it is used to ignore custom encoding charset in S3 XML body.
+func nopCharsetConverter(label string, input io.Reader) (io.Reader, error) {
+	return input, nil
+}
+
 // xmlDecoder provide decoded value in xml.
 func xmlDecoder(body io.Reader, v interface{}, size int64) error {
 	var lbody io.Reader
@@ -114,6 +120,8 @@ func xmlDecoder(body io.Reader, v interface{}, size int64) error {
 		lbody = body
 	}
 	d := xml.NewDecoder(lbody)
+	// Ignore any encoding set in the XML body
+	d.CharsetReader = nopCharsetConverter
 	return d.Decode(v)
 }
 

